LCOV_VERSION variable

sdarwin opened the pull request:
Here is a phased approach to adding a newer LCOV.  

It will default to v1.15, the same as before. The user may opt to set LCOV_VERSION, manually choosing a different version.   
 It's their responsibility to install any extra packages libcapture-tiny-perl libdatetime-perl .    An example of how to use LCOV v2.0 is included in github actions.

To consider in the future:

 codecov.sh detects if the `apt` or `yum` command is available, attempts to install perl packages, continues with an informational warning if package installation fails.

Flamefire commented on .github/workflows/ci.yml:
```diff
@@ -60,7 +60,8 @@ jobs:
               compiler: gcc-8,     cxxstd: '03,11',          os: ubuntu-20.04, install: 'g++-8-multilib', address-model: '32,64' }
-
+          - { name: Coverage lcov v2.0, coverage: yes,
+              compiler: gcc-8,     cxxstd: '03,11',          os: ubuntu-20.04, install: 'g++-8-multilib libcapture-tiny-perl libdatetime-perl', address-model: '32,64', lcov-version: 'v2.0' }
```
```suggestion
              compiler: gcc-8,     cxxstd: '03,11',          os: ubuntu-20.04, install: 'g++-8-multilib libcapture-tiny-perl', address-model: '32,64', lcov-version: 'v2.0' }
```

Wasn't that 2nd package only required for genhtml which we don't use?

sdarwin replied:
About the packages: I verified by omitting all packages, got an error, fixed the error with one package, got another error, etc.

Flamefire replied:
Ok, just wanted to double check that this isn't a debug left-over or so.

Flamefire reviewed (changes requested):
Given your comment https://github.com/boostorg/boost-ci/issues/205#issuecomment-1603298046 I'm not sure if it is really useful to a have a configurable version. As argued in that discussion I think it just has to work, the actually used version doesn't really matter. If there is reason to upgrade the version we can (usually) do so in Boost-CI `codecov.sh` globally

I specifically refer to:
> lcov: WARNING: RC option 'lcov_branch_coverage' is deprecated.  Consider using 'branch_coverage. instead.  (Backward-compatible support will be removed in the future

So a different version needs different CLI options which would make `codecov.sh` more complicated. 

However maybe we can support a few versions in that file. I.e. it currently works for 1.15 and 2.0 even if there is a warning which can be ignored. So it might be useful for experimentation and users could e.g. choose 1.16 already without any issues (I think).
And especially with the additional packages required I still think it is useful to have this opt-in

So only request I'd have for this PR is to move the LCov 2 job into old_ci.yml (maybe rename into `compatibility_ci.yml`) as Boost devs should be able to just C&P `ci.yml` and not have an extra coverage job with the same results which can be easily missed to remove. Or what do you think?

sdarwin commented:
- Moved the LCOV 2.0 test into the compatibility matrix.
- Added `--ignore-errors unused` for LCOV > v1.X  . If there was some way to fix the test for interesting files so "--ignore-errors" wasn't need, the flag could be removed again, since ideally errors shouldn't be ignored.
